### PR TITLE
Data virtual media

### PR DESCRIPTION
**Pull request by Krishnan-Priyanshu**
##### SUMMARY
migrate virtual media datasource to terraform plugin framework.

##### ISSUE TYPE

- migration pull request

##### COMPONENT NAME

- virtual media datasource
##### ADDITIONAL INFORMATION

```paste below
Running tool: /usr/local/go/bin/go test -timeout 30s -run ^TestAccRedfishVirtualMedia_fetch$ terraform-provider-redfish/redfish/provider

=== RUN   TestAccRedfishVirtualMedia_fetch
--- PASS: TestAccRedfishVirtualMedia_fetch (9.93s)
PASS
ok      terraform-provider-redfish/redfish/provider     9.956s
```
image

#### Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note

```

#### Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

#### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

### CODE REVIEW

**Comment by terraform-providers-svc** [comment]
Can one of the admins verify this patch?

File: redfish/provider/data_source_redfish_virtual_media.go [hunk 1]
```diff
@@ -1,110 +1,136 @@
...
+				Description:         "List of server BMCs and their respective user credentials",
+				Validators: []validator.List{
+					listvalidator.SizeAtMost(1),
+					listvalidator.IsRequired(),
+				},
+				NestedObject: schema.NestedBlockObject{
+					Attributes: RedfishServerDatasourceSchema(),
+				},
+			},
+			"virtual_media": schema.ListNestedBlock{
```
Comment (rounak-adhikary) [inline redfish/provider/data_source_redfish_virtual_media.go:69]: Why are we making `virtual_media` a block? What is the issue if this was an attribute of type Computed?

File: redfish/provider/resource_redfish_power.go [hunk 2]
```diff
@@ -136,10 +142,10 @@ func (r *powerResource) Create(ctx context.Context, req resource.CreateRequest,
 		return
 	}
 	// 	// Lock the mutex to avoid race conditions with other resources
-	redfishMutexKV.Lock(plan.RedfishServer.Endpoint.ValueString())
-	defer redfishMutexKV.Unlock(plan.RedfishServer.Endpoint.ValueString())
+	redfishMutexKV.Lock(plan.RedfishServer[0].Endpoint.ValueString())
+	defer redfishMutexKV.Unlock(plan.RedfishServer[0].Endpoint.ValueString())
 
-	service, err := NewConfig(r.p, &plan.RedfishServer)
+	service, err := NewConfig(r.p, &plan.RedfishServer[0])
```
Comment (rounak-adhikary) [inline redfish/provider/resource_redfish_power.go:148]: Can we push the [0] logic into the NewConfig function? Then we wont need to make changes in all ds and res

**Comment by rounak-adhikary** [comment]
Please enable linting for ur datasource

File: redfish/provider/resource_redfish_power.go (hunk 2 above)
Comment (rounak-adhikary) [reply redfish/provider/resource_redfish_power.go:148]: Better yet, let's push the locking logic as well to NewConfig.

Comment (Krishnan-Priyanshu) [reply redfish/provider/resource_redfish_power.go]: pushed the [0] logic to NewConfig Function

File: redfish/provider/common.go [hunk 3]
```diff
@@ -88,19 +88,24 @@ func getSystemResource(service *gofish.Service) (*redfish.ComputerSystem, error)
 // NewConfig function creates the needed gofish structs to query the redfish API
 // See https://github.com/stmcginnis/gofish for details. This function returns a Service struct which can then be
 // used to make any required API calls.
-func NewConfig(pconfig *redfishProvider, rserver *models.RedfishServer) (*gofish.Service, error) {
+func NewConfig(pconfig *redfishProvider, rserver *[]models.RedfishServer) (*gofish.Service, error) {
+	if len(*rserver) < 1 {
```
Comment (rounak-adhikary) [inline redfish/provider/common.go:92]: Looks like this logic can be added in redfish server schema validator https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework-validators@v0.12.0/listvalidator#SizeBetween

Comment (Krishnan-Priyanshu) [reply redfish/provider/common.go:92]: we've added that validation too at schema level.

Comment (rounak-adhikary) [reply redfish/provider/common.go:92]: Thats not true. The schema validator says that no more than 1 server details can be provided, whereas the highlighted line of code validates that no less than 1 server detail is provided.

Comment (Krishnan-Priyanshu) [reply redfish/provider/common.go:92]: removing the code
To-Do: Verify from plan modifier, if required implement wrapper for validation of unknown in redfish_server.

**Review by rounak-adhikary — APPROVED** [review]

**Review by Sakshi-dell — APPROVED** [review]

**Pull request merged**